File: cmd/pmd-gen/read.go
package main

import (
	"github.com/Karthik99999/pmd-gen/internal/romdata"
	"github.com/Karthik99999/pmd-gen/internal/utils"
	"strings"
)

type rescueCode struct {
	code []string
}

// NewRescueCode returns a new instance of the rescueCode struct
func NewRescueCode(code string) *rescueCode {
	rescueCode := &rescueCode{}
	code = strings.ReplaceAll(code, " ", "")
	var codef string
	for i := 0; i < len(code); i += 2 {
		codef += string(code[i:i+2] + " ")
	}
	codef = strings.TrimSpace(codef)
	rescueCode.code = strings.Split(codef, " ")
	return rescueCode
}

// unshuffled password symbols
func (rc *rescueCode) unshuffle() []string {
	unshuffledIndex := []int{3, 27, 13, 21, 12, 9, 7, 4, 6, 17, 19, 16, 28, 29, 23, 20, 11, 0, 1, 22, 24, 14, 8, 2, 15, 25, 10, 5, 18, 26}
	unshuffled := make([]string, 30)
	for i := 0; i < len(unshuffledIndex); i++ {
		unshuffled[i] = rc.code[unshuffledIndex[i]]
	}
	return unshuffled
}

// convert symbols to indexes 0-63
// xs is not used in passwords due to 64 being a 7 bit number
func (rc *rescueCode) toIndexes() []int {
	symbols := []string{"1f", "2f", "3f", "4f", "5f", "6f", "7f", "8f", "9f", "pf", "mf", "df", "xf",
		"1h", "2h", "3h", "4h", "5h", "6h", "7h", "8h", "9h", "ph", "mh", "dh", "xh",
		"1w", "2w", "3w", "4w", "5w", "6w", "7w", "8w", "9w", "pw", "mw", "dw", "xw",
		"1e", "2e", "3e", "4e", "5e", "6e", "7e", "8e", "9e", "pe", "me", "de", "xe",
		"1s", "2s", "3s", "4s", "5s", "6s", "7s", "8s", "9s", "ps", "ms", "ds"}
	indexes := make([]int, 30)
	unshuffled := rc.unshuffle()
	for i := 0; i < len(unshuffled); i++ {
		for j := 0; j < len(symbols); j++ {
			if symbols[j] == strings.ToLower(unshuffled[i]) {
				indexes[i] = j
			}
		}
	}
	return indexes
}

// bitpack the code
func (rc *rescueCode) bitpack() []int {
	var newcode []int
	reader := utils.NewBitstreamReader(rc.toIndexes(), 6)
	for reader.Remaining() {
		newcode = append(newcode, reader.Read(8))
	}
	return newcode
}

// decrypt the code by using the same rng method used to encrypt
func (rc *rescueCode) decrypt() []int {
	bitpacked := rc.bitpack()
	newcode := []int{bitpacked[0], bitpacked[1]}
	seed := bitpacked[0] | bitpacked[1]<<8
	rng := utils.NewRNG(seed)
	for i := 2; i < len(bitpacked); i++ {
		rand := rng.NextInt()
		newcode = append(newcode, (bitpacked[i]-rand)&0xFF)
	}
	remain := 8 - (len(bitpacked) * 8 % 6)
	newcode[len(newcode)-1] &= (1 << remain) - 1
	return newcode
}

// calculate checksum
func checksum(code []int) int {
	var sum int
	for _, b := range code {
		sum += b
	}
	checksum := ^(sum + (sum >> 8)) & 0xFF
	return checksum
}

func crc32(bytes string) int {
	sum := 0xFFFFFFFF
	for i := 0; i < len(bytes); i++ {
		sum = romdata.GetRomData().Crc32Table[(sum&0xFF)^int(bytes[i])] ^ (sum >> 8)
	}
	return sum ^ 0xFFFFFFFF
}

type rescueInfo struct {
	InclChecksum int   `json:"incl_checksum"`
	CalcChecksum int   `json:"calc_checksum"`
	Timestamp    int   `json:"timestamp"`
	Type         int   `json:"type"`
	Unknown1     int   `json:"unknown1"`
	Team         []int `json:"team"`
	Dungeon      int   `json:"dungeon"`
	Floor        int   `json:"floor"`
	Pokemon      int   `json:"pokemon"`
	Gender       int   `json:"gender"`
	Reward       int   `json:"reward"`
	Unknown2     int   `json:"unknown2"`
	Revive       int   `json:"revive"`
}

func (rc *rescueCode) deserialize() *rescueInfo {
	code := rc.decrypt()
	info := &rescueInfo{
		InclChecksum: code[0],
		CalcChecksum: checksum(code[1:]),
	}

	reader := utils.NewBitstreamReader(code[1:], 8)
	info.Timestamp = reader.Read(32)
	info.Type = reader.Read(1)
	info.Unknown1 = reader.Read(1)
	var team []int
	for i := 0; i < 12; i++ {
		team = append(team, reader.Read(9))
	}
	info.Team = team
	if info.Type == 0 {
		info.Dungeon = reader.Read(7)
		info.Floor = reader.Read(7)
		info.Pokemon = reader.Read(11)
		info.Gender = reader.Read(2)
		info.Reward = reader.Read(2)
		info.Unknown2 = reader.Read(1)

		symbols := []string{"1f", "2f", "3f", "4f", "5f", "6f", "7f", "8f", "9f", "pf", "mf", "df", "xf",
			"1h", "2h", "3h", "4h", "5h", "6h", "7h", "8h", "9h", "ph", "mh", "dh", "xh",
			"1w", "2w", "3w", "4w", "5w", "6w", "7w", "8w", "9w", "pw", "mw", "dw", "xw",
			"1e", "2e", "3e", "4e", "5e", "6e", "7e", "8e", "9e", "pe", "me", "de", "xe",
			"1s", "2s", "3s", "4s", "5s", "6s", "7s", "8s", "9s", "ps", "ms", "ds"}
		indexes := make([]int, 30)
		for i := 0; i < len(rc.code); i++ {
			for j := 0; j < len(symbols); j++ {
				if symbols[j] == strings.ToLower(rc.code[i]) {
					indexes[i] = j
				}
			}
		}
		charcode := ""
		for _, x := range indexes {
			charcode += string([]rune(romdata.GetRomData().Charmap)[x])
		}
		info.Revive = crc32(charcode) & 0x3FFFFFFF
	} else {
		info.Revive = reader.Read(30)
	}

	return info
}

File: src/lib/rtdx/read.ts
import { BitstreamReader } from '../bitstream';
import { checksum, crc32, RNG, symbols } from './utils';
import { Data } from './data';

/**
 * Splits code string into array
 */
function splitCode(code: string): string[] {
	code = code.replace(' ', '').toLowerCase();
	let codeToSplit = '';
	for (let i = 0; i < code.length; i += 2) {
		codeToSplit += code.slice(i, i+2) + ' ';
	}
	return codeToSplit.trim().split(' ');
}

/** 
 * Unshuffles the password symbols
 */
function unshuffle(code: string[]): string[] {
	const unshuffledIndex = [
        3, 27, 13, 21, 12, 9, 7, 4, 6, 17, 19, 16, 28, 29, 23, 20, 11, 0, 1, 22, 24, 14, 8, 2, 15, 25, 10, 5, 18, 26,
    ];
	const unshuffled: string[] = [];
	for (let i = 0; i < 30; i++) {
		unshuffled[i] = code[unshuffledIndex[i]];
	}
	return unshuffled;
}

/**
 * Converts symbols to indexes 0-63
 */
function toIndexes(code: string[]): number[] {
	const indexes: number[] = [];
	for (let i = 0; i < code.length; i++) {
		for (let j = 0; j < symbols.length; j++) {
			if (symbols[j] === code[i]) {
				indexes[i] = j;
			}
		}
	}
	return indexes;
}

/**
 * Bitpacks the code
 */
function bitpack(indexes: number[]): number[] {
	const newCode: number[] = [];
	const reader = new BitstreamReader(indexes, 6);
	while (reader.remaining()) {
		newCode.push(reader.read(8));
	}
	return newCode;
}

/**
 * Decrypt the code by using the same rng method used to encrypt
 */
function decrypt(code: number[]): number[] {
	const newcode = code.slice(0, 2);
	const seed = code[0] | code[1] << 8;
	const rng = new RNG(seed);
	for (const c of code.slice(2)) {
		const rand = rng.next();
		newcode.push((c - rand) & 0xFF);
	}
	const remain = 8 - (code.length * 8 % 6);
	newcode[newcode.length - 1] &= (1 << remain) - 1;
	return newcode;
}

/**
 * Data structure of a password
 */
interface PasswordData {
	password: string;
	inclChecksum: number;
	calcChecksum: number;
	timestamp: number;
	unknown1: number;
	team: number[];
	dungeon: number;
	floor: number;
	pokemon: number;
	gender: number;
	reward: number;
	unknown2: number;
	revive: number;
}

type RescueData = PasswordData & {type: 0};
type RevivalData = Omit<PasswordData, 'dungeon' | 'floor' | 'pokemon' | 'gender' | 'reward' | 'unknown2'> & {type: 1};

/**
 * Deserializes a password and returns relevant data
 * for internal use. Use read() to get human-readable data.
 * @param password Password as a string of letters and numbers
 * to represent the icons from the game. The letter or number
 * on the icon should come first, then a letter represnting the
 * shape on the icon should come right after.
 * 
 * Fire - F
 * 
 * Heart - H
 * 
 * Water Drop - W
 * 
 * Emerald - E
 * 
 * Star - S
 * 
 * For example, a 1 on top of a water drop should be 1W.
 */
function deserialize(password: string): RescueData | RevivalData {
    const passwordArr = splitCode(password);
	const unshuffled = unshuffle(passwordArr);
	const indexes = toIndexes(unshuffled);
	const bitpacked = bitpack(indexes);
	const code = decrypt(bitpacked);

	const inclChecksum = code[0];
	const calcChecksum = checksum(code.slice(1));

	const reader = new BitstreamReader(code.slice(1), 8);
	const timestamp = reader.read(32);
	const type = reader.read(1);
	const unknown1 = reader.read(1);
	const team: number[] = [];
	for (let i = 0; i < 12; i++) {
		team.push(reader.read(9));
	}
	if (type === 0) {
		const dungeon = reader.read(7);
		const floor = reader.read(7);
		const pokemon = reader.read(11);
		const gender = reader.read(2);
		const reward = reader.read(2);
		const unknown2 = reader.read(1);

		const indexes = toIndexes(passwordArr);
		let charcode = '';
		for (const i of indexes) {
			charcode += Data.charmap[i];
		}
		const revive = crc32(charcode) & 0x3FFFFFFF;

		const data: RescueData = {
			password,
			inclChecksum,
			calcChecksum,
			timestamp,
			type,
			unknown1,
			team,
			dungeon,
			floor,
			pokemon,
			gender,
			reward,
			unknown2,
			revive,
		};
		return data;
	} else {
		const revive = reader.read(30);

		const data: RevivalData = {
			password,
			inclChecksum,
			calcChecksum,
			timestamp,
			type: 1,
			unknown1,
			team,
			revive,
		};
		return data;
	}
}

/**
 * Human-friendly data structure of rescue passwords
 */
interface PasswordInfo {
	password: string;
	valid: boolean;
	timestamp: number;
	teamName: string;
	dungeon: string;
	floor: string;
	pokemon: string;
	gender: 'Male' | 'Female' | 'Genderless';
	reward: 'Deluxe' | 'Special' | 'Regular';
	revive: number;
}

type RescueInfo = PasswordInfo & {type: 'Rescue'};
type RevivalInfo = Omit<PasswordInfo, 'dungeon' | 'floor' | 'pokemon' | 'gender' | 'reward'> & {type: 'Revival'};

/**
 * Parses the raw password data and returns relevant human-readable data
 */
function parseData(data: RescueData | RevivalData): RescueInfo | RevivalInfo {
	const valid = (data.inclChecksum === data.calcChecksum);
	let teamName = '';
	for (const i of data.team) {
		if (i === 0) break;
		if (i < 402) {
			teamName += Data.charmap_text[i];
		} else {
			teamName += '★';
		}
	}
	if (data.type === 0) {
		let dungeon = '';
		let floor = '';
		let pokemon = '';
		const dungeonData = Data.dungeons.get(data.dungeon);
		if (dungeonData.valid) {
			dungeon = dungeonData.name;
			floor = `${dungeonData.ascending ? '' : 'B'}${data.floor}F`;
		}
		const pokemonData = Data.pokemon.get(data.pokemon);
		if (pokemonData.valid) {
			pokemon = `${pokemonData.name}${pokemonData.const.startsWith('YOBI') ? ' (Shiny)' : ''}`;
		}
		const gender = data.gender === 0 ? 'Male' : data.gender === 1 ? 'Female' : 'Genderless';
		const reward = data.reward === 3 ? 'Deluxe' : data.reward === 2 ? 'Special' : 'Regular';

		const info: RescueInfo = {
			password: data.password,
			valid,
			timestamp: data.timestamp,
			type: 'Rescue',
			teamName,
			dungeon,
			floor,
			pokemon,
			gender,
			reward,
			revive: data.revive,
		};
		return info;
	} else if (data.type === 1) {
		const info: RevivalInfo = {
			password: data.password,
			valid,
			timestamp: data.timestamp,
			type: 'Revival',
			teamName,
			revive: data.revive,
		};
		return info;
	} else {
		// Empty data
		return {
			password: '',
			valid: false,
			timestamp: 0,
			type: 'Revival',
			teamName: '',
			revive: 0,
		};
	}
}
/**
 * Reads a password and returns human-readable data
 * @param password Password as a string of letters and numbers
 * to represent the icons from the game. The letter or number
 * on the icon should come first, then a letter represnting the
 * shape on the icon should come right after.
 * 
 * Fire - F
 * 
 * Heart - H
 * 
 * Water Drop - W
 * 
 * Emerald - E
 * 
 * Star - S
 * 
 * For example, a 1 on top of a water drop should be 1W.
 */
export function read(password: string): RescueInfo | RevivalInfo {
	const data = deserialize(password);
	const info = parseData(data);
	return info;
}
